### ISSUE TITLE
📝 HTML: `Invalid document: Expected end tag of kind Fill but found Indent.`

### ISSUE THREAD

**Issue by globau** [biomejs/biome#6623]
### Environment information

```bash
CLI:
  Version:                      2.0.6
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v24.3.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         npm/11.4.2

Biome Configuration:
  Status:                       Not set
  Path:                         unset

Workspace:
  Open Documents:               0
```

### Configuration

```JSON
{
  "root": false,
  "$schema": "https://biomejs.dev/schemas/2.0.0/schema.json",
  "formatter": {
    "enabled": true,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineWidth": 88
  },
  "assist": {
    "enabled": true,
    "actions": {
      "source": {
        "organizeImports": "on",
        "useSortedProperties": "on"
      }
    }
  },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "style": {
        "noDescendingSpecificity": "off",
        "noInferrableTypes": "error",
        "noParameterAssign": "off",
        "noUnusedTemplateLiteral": "error",
        "noUselessElse": "error",
        "useAsConstAssertion": "error",
        "useDefaultParameterLast": "error",
        "useEnumInitializers": "error",
        "useNumberNamespace": "error",
        "useSelfClosingElements": "error",
        "useSingleVarDeclarator": "error"
      },
      "correctness": {
        "noUnusedImports": "error"
      },
      "suspicious": {
        "noConsole": "error"
      }
    }
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "double"
    }
  },
  "html": {
    "formatter": {
      "enabled": true
    }
  }
}
```

### Playground link

https://biomejs.dev/playground/?indentStyle=space&indentWidth=4&bracketSameLine=true&indentScriptAndStyle=true&code=PAAhAGQAbwBjAHQAeQBwAGUAIABoAHQAbQBsAD4ACgA8AGgAdABtAGwAPgAKADwAYgBvAGQAeQA%2BAAoAPABiAD4APAAvAGIAPgAgACYAbQBkAGEAcwBoADsACgA8AGIAPgA8AC8AYgA%2BAAoAPAAvAGIAbwBkAHkAPgAKADwALwBoAHQAbQBsAD4A&language=html

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by dyc3** [comment]
Man, I really have been picking the hard issues lately...

I'm pretty confident that this is `FormatHtmlElementList`'s fault in some way (it usually is when it comes to the HTML formatter). In fact, I can point to the write_elements calls that cause this, because removing the start/end fill tags makes the panic go away:

https://github.com/biomejs/biome/blob/e36808f01bb529612ebca217df11b9e8c2c8ef46/crates/biome_html_formatter/src/html/lists/element_list.rs#L743-L747

Even though the element list formatter does things at a pretty low level like this, I haven't yet found a place where we actually mishandle these tags, like forgetting to add an end tag or something.

The exact indentation tags that its conflicting with are being emitted here:

https://github.com/biomejs/biome/blob/e36808f01bb529612ebca217df11b9e8c2c8ef46/crates/biome_html_formatter/src/html/lists/element_list.rs#L790

So essentially, what _should_ be emitted:

```
StartIndent
    StartFill
         content...
    EndFill
EndIndent
```

However, when processing the print queue, it throws the error when it tries to consume the `StartFill`. It doesn't even get to the end tags to know if its a mismatch. 

It's also not `print_fill_entries` itself that raises the error, its gotta be something further down the callstack.

I'll come back to this after lunch.

**Comment by dyc3** [comment]
I've found this error gets raised in `fits_element`, which is used to measure format elements to see if they fit on the line for "fill" formatting.

So far I've only been able to glean that `FitsCallStack` does not get enough `Fill`s popped off the stack before it tries to pop the `Indent`, which causes the error.

<details>

<summary>This is what the `FitsCallStack` looks like during execution, up to the point where the error occurs.</summary>

```
|   |   |   |   Pushing tag: Fill
|   |   |   |   |   Pushing tag: Fill
|   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   |   Pushing tag: Group
|   |   |   |   |   |   |   Pop tag: Group
|   |   |   |   |   |   |   Pushing tag: ConditionalContent
|   |   |   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   |   |   Pop tag: ConditionalContent
|   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   Pop tag: Fill
|   |   |   |   |   Pushing tag: Fill
|   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   Pop tag: Fill
|   |   |   |   |   Pushing tag: Fill
|   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   Pushing tag: Group
|   |   |   |   |   |   Pop tag: Group
|   |   |   |   |   |   Pushing tag: ConditionalContent
|   |   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   |   Pop tag: ConditionalContent
|   |   |   |   |   Pop tag: Entry
|   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   Pop tag: Entry
|   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   Pop tag: Entry
|   |   |   |   |   Pushing tag: Fill
|   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   |   Pushing tag: Group
|   |   |   |   |   |   |   Pop tag: Group
|   |   |   |   |   |   |   Pushing tag: ConditionalContent
|   |   |   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   |   |   Pop tag: ConditionalContent
|   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   |   Pushing tag: Entry
|   |   |   |   |   |   Pop tag: Entry
|   |   |   |   |   Pop tag: Fill
|   |   |   |   Pop tag: Indent
```

</details>

I haven't been making any progress on this, so I'll unassign myself for now.

**Comment by sebhildebrandt** [comment]
I am getting the same error her on my side in one of our HTML files (all others work fine)... would it help, if you get this specific HTML code?

**Comment by dyc3** [comment]
@sebhildebrandt We already have a minimal reproduction.

**Comment by mj026** [comment]
In case you still need a reproduction.

Formatting works with this html: 
```html
<div>
  This website is built with
  <a href="https://tailwindcss.com/">tailwindcss</a>
  <a href="https://vite.dev/">vite</a>
  <a href="https://biomejs.dev/">biome</a>
</div>
```

When a trailing `,` is added to one of the links, it will say `Invalid document: Expected end tag of kind Fill but found Indent`
```html
<div>
  This website is built with
  <a href="https://tailwindcss.com/">tailwindcss</a>,
  <a href="https://vite.dev/">vite</a>
  <a href="https://biomejs.dev/">biome</a>
</div>

**Comment by dyc3** [comment]
This is fixed on `next`, and will be released with 2.4
